src/hash.c

```diff
@@ -41,6 +41,59 @@ static uint32_t hash_double(double x) {
   return hash_int64(value.i);
 }
 
+static uint32_t hash_char(SEXP x) {
+  return hash_int64((int64_t) x);
+}
+
+static uint32_t hash_char_translate(SEXP x) {
+  const void *vmax = vmaxget();
+
+  const char* x_utf8 = CHAR_IS_UTF8(x) ? CHAR(x) : Rf_translateCharUTF8(x);
+
+  uint32_t hash = 0;
+
+  while(*x_utf8++) {
+    hash = hash_combine(hash, hash_int64((int64_t) *x_utf8));
+  }
+
+  vmaxset(vmax);
+  return hash;
+}
+
+// Determine if `x` must be translated to UTF-8
+// - CHAR_IS_BYTES - Don't translate these cases:
+//   (bytes + bytes), (bytes + unknown), (bytes + utf8), (bytes + latin1)
+// - CHAR_ENC_TYPE - Don't translate these cases:
+//   (utf8 + utf8), (latin1 + latin1), (unknown + unknown)
+// - Translate these:
+//   (utf8 + latin1), (unknown + utf8), (unknown + latin1)
+static bool chr_requires_translation(SEXP x, R_len_t size) {
+  const SEXP* xp = STRING_PTR_RO(x);
+  SEXP xp_val;
+
+  bool translate = false;
+  int first_encoding = CHAR_ENC_TYPE(*xp);
+
+  for (R_len_t i = 0; i < size; ++i, ++xp) {
+    xp_val = *xp;
+
+    if (CHAR_IS_BYTES(xp_val)) {
```

Comment:
So if any element of the vector has bytes encoding, we don't translate?

Reply:
Correct, that's an error with base R (b/c its an error with `Rf_translateCharUTF8()`), see the last example with `vec_equal` below

Reply:
I think the only disadvantage is that it means that you can't short-circuit and return `true` as soon as you see an encoding that's different to the first.

Reply:
That's true. Luckily if you do need to short circuit and return `true`, then that means you need to alter encodings, which takes _much_ longer than this loop, even if it has to cycle through all the values.